pkg/cloud/gcpprovider.go

```diff
@@ -1349,12 +1356,27 @@ func (gcp *GCP) AllNodePricing() (interface{}, error) {
 // NodePricing returns GCP pricing data for a single node
 func (gcp *GCP) NodePricing(key Key) (*Node, error) {
 	gcp.DownloadPricingDataLock.RLock()
-	defer gcp.DownloadPricingDataLock.RUnlock()
 	if n, ok := gcp.Pricing[key.Features()]; ok {
 		klog.V(4).Infof("Returning pricing for node %s: %+v from SKU %s", key, n.Node, n.Name)
 		n.Node.BaseCPUPrice = gcp.BaseCPUPrice
+		gcp.DownloadPricingDataLock.RUnlock()
 		return n.Node, nil
+	} else if _, ok := gcp.ValidPricingKeys[key.Features()]; ok {
+		gcp.DownloadPricingDataLock.RUnlock()
+		err := gcp.DownloadPricingData()
+		if err != nil {
+			return nil, fmt.Errorf("Download pricing data failed: %s", err.Error())
+		}
+		if n, ok := gcp.Pricing[key.Features()]; ok {
+			klog.V(4).Infof("Returning pricing for node %s: %+v from SKU %s", key, n.Node, n.Name)
+			n.Node.BaseCPUPrice = gcp.BaseCPUPrice
+			return n.Node, nil
+		}
+		klog.V(1).Infof("[Warning] no pricing data found for %s: %s", key.Features(), key)
+		gcp.DownloadPricingDataLock.RUnlock()
```

Comment:
This seems unnecessary maybe. At this point, we would've already release the read lock. However, if you do something like the suggestion above, this line would need to be `gcp.DownloadPricingDataLock.Unlock()`

The other possibility for ALL of this code is create a lock-less `downloadPricing()` defined in the way described above, and then just acquire the read/write lock and defer it at the top of the method. It doesn't leverage the most optimal RW locking, but the constant lock switching probably creates more performance impact than just a single read/write lock on the entire method.